src/payment/models/buy-crypto/entities/buy-crypto-batch.entity.ts

```diff
@@ -0,0 +1,192 @@
+
+    return this;
+  }
+
+  complete(): this {
+    this.status = BuyCryptoBatchStatus.COMPLETE;
+
+    return this;
+  }
+
+  pending(purchaseTxId: string): this {
+    this.purchaseTxId = purchaseTxId;
+    this.status = BuyCryptoBatchStatus.PENDING_LIQUIDITY;
+
+    return this;
+  }
+
+  payingOut(): this {
+    this.status = BuyCryptoBatchStatus.PAYING_OUT;
+
+    return this;
+  }
+
+  recordDexToOutTransfer(txId: string): this {
+    this.outTxId = txId;
+
+    return this;
+  }
+
+  groupPayoutTransactions(): BuyCrypto[][] {
+    if (this.status !== BuyCryptoBatchStatus.SECURED) {
+      throw new Error(`Cannot payout batch which is not secured. Batch ID: ${this.id}. Batch status: ${this.status}`);
+    }
+
+    console.info(`Grouping transactions for payout. Batch ID: ${this.id}`);
+
+    const payoutTransactions = this.transactions.filter((tx) => !tx.txId);
+
+    if (this.transactions.length > 0 && payoutTransactions.length !== this.transactions.length) {
+      console.warn(
+        `Skipped ${this.transactions.length - payoutTransactions.length} transactions of batch ID: ${
+          this.id
+        } to avoid double payout.`,
+      );
+    }
+
+    const maxGroupSize = this.outputAsset === 'DFI' ? 100 : 10;
+    const groups = this.createPayoutGroups(payoutTransactions, maxGroupSize);
+
+    console.info(`Created ${groups.length} transaction group(s) for payout. Batch ID: ${this.id}`);
+
+    return groups;
+  }
+
+  private createPayoutGroups(transactions: BuyCrypto[], maxGroupSize: number): BuyCrypto[][] {
+    const result: Map<number, BuyCrypto[]> = new Map();
+
+    let currentGroupNumber = 0;
+
+    transactions.forEach((tx) => {
+      let currentGroup = result.get(currentGroupNumber);
+
+      if (!currentGroup) {
+        currentGroup = [tx];
+        result.set(currentGroupNumber, currentGroup);
+
+        return;
+      }
+
+      if (currentGroup.find((_tx) => _tx.buy.user.address === tx.buy.user.address)) {
+        // find nearest non-full group without repeating address
+        const suitableExistingGroups = [...result.entries()].filter(
+          // eslint-disable-next-line @typescript-eslint/no-unused-vars
+          ([_, transactions]) =>
+            transactions.length < maxGroupSize &&
+            !transactions.find((_tx) => _tx.buy.user.address === tx.buy.user.address),
+        );
+
+        if (suitableExistingGroups.length) {
+          const [key, group] = suitableExistingGroups[0];
+          result.set(key, [...group, tx]);
+
+          return;
+        }
+
+        const newGroup = [tx];
+        result.set(result.size + 1, newGroup);
+
+        return;
+      }
+
+      if (currentGroup.length >= maxGroupSize) {
+        const newGroup = [tx];
+        result.set(currentGroupNumber + 1, newGroup);
+        currentGroupNumber++;
+
+        return;
+      }
+
+      result.set(currentGroupNumber, [...currentGroup, tx]);
+    });
+
+    return [...result.values()];
+  }
```

Comment:
Easier solution?

```ts
  private createPayoutGroups(transactions: BuyCrypto[], maxGroupSize: number): BuyCrypto[][] {
    const result: Map<number, BuyCrypto[]> = new Map();

    transactions.forEach((tx) => {
      // find nearest non-full group without repeating address
      const suitableExistingGroups = [...result.entries()].filter(
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        ([_, transactions]) =>
          transactions.length < maxGroupSize &&
          !transactions.find((_tx) => _tx.buy.user.address === tx.buy.user.address),
      );

      const [key, group] = suitableExistingGroups[0] ?? [result.size, []];
      result.set(key, [...group, tx]);
    });

    return [...result.values()];
  }
```
Disclaimer: no guarantee of correctness ... I did test it though ;-)

Reply:
very nice! thanks! unit tests pass :)